sympy/functions/combinatorial/factorials.py

```diff
@@ -1029,12 +1064,50 @@ def _eval_expand_func(self, **hints):
         else:
             return binomial(*self.args)
 
+    def _eval_rewrite_as_Piecewise(self, n, k, **kwargs):
+        from sympy.concrete.products import Product
+        from sympy import gamma, Abs, nan
+        knk = lambda x: Or(Eq(k, x), Eq(n - k, x))
+        isint = lambda x: Eq(x, floor(x))
+        nint = lambda x: Ne(x, floor(x))
+        pos = lambda x: And(Ne(x, 0), Eq(x/Abs(x), 1))
+        neg = lambda x: And(Ne(x, 0), Eq(x/Abs(x), -1))
+        inf = lambda x: And(Ne(x, 0), Eq(1/x, 0))
+        for i in igen():
+            if not n.has(i) and not k.has(i):
+                break
+        return Piecewise(
+            (1, Or(Eq(n, k), knk(0))),
+            (n, knk(1)),
+            (zoo, And(neg(n), isint(n), nint(k))),
+            (nan, Or(inf(n), inf(k))),
+            (0, Or(
+                knk(-k),
+                And(knk(-1), Or(nint(n), Ne(n, -1))),
+                And(
+                    And(isint(n), isint(k)),
+                    Or(And(pos(n*k), pos(k - n)),
+                        And(pos(n), neg(k)))))
```

Comment:
`pos(n)` should be `Or(Eq(n, 0), pos(n))` since the condition is `n.is_nonnegative`